Mathlib/RingTheory/Regular/RegularSequence.lean

```diff
@@ -0,0 +1,645 @@
+    ofList (r :: rs) • N = r • N ⊔ ofList rs • N := by
+  rw [ofList_cons, Submodule.sup_smul, Submodule.ideal_span_singleton_smul]
+
+end Ideal
+
+namespace Submodule
+
+lemma smul_top_le_comap_smul_top [CommSemiring R] [AddCommMonoid M]
+    [Module R M] [AddCommMonoid N] [Module R N] (I : Ideal R) (f : M →ₗ[R] N) :
+    I • ⊤ ≤ comap f (I • ⊤) :=
+  map_le_iff_le_comap.mp <| le_of_eq_of_le (map_smul'' _ _ _) <|
+    smul_mono_right _ le_top
+
+variable (M) [CommRing R] [AddCommGroup M] [Module R M] [AddCommGroup N]
+    [Module R N] (r : R) (rs : List R)
+
+/-- The equivalence between M ⧸ (r₀, r₁, …, rₙ)M and (M ⧸ r₀M) ⧸ (r₁, …, rₙ) (M ⧸ r₀M). -/
+def quotOfListConsSMulTopEquivQuotSMulTopInner :
+    (M ⧸ (Ideal.ofList (r :: rs) • ⊤ : Submodule R M)) ≃ₗ[R]
+      QuotSMulTop r M ⧸ (Ideal.ofList rs • ⊤ : Submodule R (QuotSMulTop r M)) :=
+  quotEquivOfEq _ _ (Ideal.ofList_cons_smul r rs ⊤) ≪≫ₗ
+    (quotientQuotientEquivQuotientSup (r • ⊤) (Ideal.ofList rs • ⊤)).symm ≪≫ₗ
+      quotEquivOfEq _ _ (by rw [map_smul'', map_top, range_mkQ])
+
+/-- The equivalence between M ⧸ (r₀, r₁, …, rₙ)M and (M ⧸ (r₁, …, rₙ)) ⧸ r₀ (M ⧸ (r₁, …, rₙ)). -/
+def quotOfListConsSMulTopEquivQuotSMulTopOuter :
+    (M ⧸ (Ideal.ofList (r :: rs) • ⊤ : Submodule R M)) ≃ₗ[R]
+      QuotSMulTop r (M ⧸ (Ideal.ofList rs • ⊤ : Submodule R M)) :=
+  quotEquivOfEq _ _ (Eq.trans (Ideal.ofList_cons_smul r rs ⊤) (sup_comm _ _)) ≪≫ₗ
+    (quotientQuotientEquivQuotientSup (Ideal.ofList rs • ⊤) (r • ⊤)).symm ≪≫ₗ
+      quotEquivOfEq _ _ (by rw [map_pointwise_smul, map_top, range_mkQ])
+
+variable {M}
+
+lemma quotOfListConsSMulTopEquivQuotSMulTopInner_naturality (f : M →ₗ[R] N) :
+    (quotOfListConsSMulTopEquivQuotSMulTopInner N r rs).toLinearMap ∘ₗ
+        mapQ _ _ _ (smul_top_le_comap_smul_top (Ideal.ofList (r :: rs)) f) =
+      mapQ _ _ _ (smul_top_le_comap_smul_top _ (QuotSMulTop.map r f)) ∘ₗ
+        (quotOfListConsSMulTopEquivQuotSMulTopInner M r rs).toLinearMap :=
+  quot_hom_ext _ _ _ fun _ => rfl
+
+lemma top_eq_ofList_cons_smul_iff :
+    (⊤ : Submodule R M) = Ideal.ofList (r :: rs) • ⊤ ↔
+      (⊤ : Submodule R (QuotSMulTop r M)) = Ideal.ofList rs • ⊤ := by
+  conv => congr <;> rw [eq_comm, ← subsingleton_quotient_iff_eq_top]
+  exact (quotOfListConsSMulTopEquivQuotSMulTopInner M r rs).toEquiv.subsingleton_congr
+
+end Submodule
+
+namespace RingTheory.Sequence
+
+open scoped TensorProduct List
+open Function Submodule QuotSMulTop
+
+section Definitions
+
+-- In theory, regularity of `rs : List α` on `M` makes sense as soon as
+-- `[Monoid α]`, `[AddCommGroup M]`, and `[DistribMulAction α M]`.
+-- Instead of `Ideal.ofList (rs.take i) • (⊤ : Submodule R M)` we use
+-- `⨆ (j : Fin i), rs[j] • (⊤ : AddSubgroup M)`.
+-- However it's not clear that this is a useful generalization.
+-- If we add the assumption `[SMulCommClass α α M]` this is essentially the
+-- same as focusing on the commutative ring case, passing to the monoid ring
+-- `ℤ[abelianization of α]`.
```

Comment:
Can/should this be a module docstring? If not, please drop the `!` on the first line of this suggestion.
```suggestion
/-!
In theory, regularity of `rs : List α` on `M` makes sense as soon as
`[Monoid α]`, `[AddCommGroup M]`, and `[DistribMulAction α M]`.
Instead of `Ideal.ofList (rs.take i) • (⊤ : Submodule R M)` we use
`⨆ (j : Fin i), rs[j] • (⊤ : AddSubgroup M)`.
However it's not clear that this is a useful generalization.
If we add the assumption `[SMulCommClass α α M]` this is essentially the
same as focusing on the commutative ring case, passing to the monoid ring
`ℤ[abelianization of α]`.
-/
```

Reply:
Hmm, not sure it makes sense as a module docstring. I wanted to include it as a warning/suggestion(?) to future maintainers of this code, explaining why I defined it like this and not in another plausible way. I don't think it's useful for a client looking at the documentation to see it. I've changed to a multi-line comment though